### DISCUSSION TITLE
How can I know if my ApnsClient is not able to cope with the amount of notifications I'm sending through it?

### DISCUSSION THREAD

**Discussion by michaeldiener** [relayrides/pushy#956] [Q&A] [answered]
Hi there!

How can I know if my ApnsClient is not able to cope with the amount of notifications I'm sending through it?
Would there be an exception thrown somewhere or can I check the amount of queued up notifications?

Thanks,
Michael

**Comment by jchambers** [comment, marked as answer]
Well, to answer your question directly, you can measure the number of in-flight messages with a [metrics listener](https://pushy-apns.org/apidocs/latest/com/eatthepath/pushy/apns/ApnsClientMetricsListener.html); you can figure out how many messages are in flight by measuring the difference between the number of notifications sent and the number accepted/rejected by the server.

That said, if your goal is to actually _control_ how many messages are in flight, I recommend implementing flow control outside of Pushy; please see the [flow control section of the "best practices" wiki page](https://github.com/jchambers/pushy/wiki/Best-practices#flow-control) for details.

**Reply by michaeldiener** [reply]
Thank you! I meant something slightly different, but the metrics listener will also work well for this.
Just for clarification, I want to count the messages handed over to the send method, but which haven't actually been sent yet. Your idea will also work well for this, thanks for your guidance and not sure why I didn't think about this before!
